gnomad/variant_qc/evaluation.py

```diff
@@ -116,108 +74,78 @@ def quantiles_to_bin_boundaries(quantiles: List[int]) -> Dict:
-                        ),
-                    ),
-                )
-                for bin_id in bin_expr
-            }
-        )
-    )
-
-    # Take care of bins with duplicated boundaries
-    bin_stats = bin_stats.annotate(
+    bin_ht = bin_ht.order_by("_score", "_rand")
+    bin_ht = bin_ht.annotate(
         **{
-            rname: bin_stats[rname].annotate(
-                **quantiles_to_bin_boundaries(bin_stats[rname].quantiles)
+            f"{bin_id}_rank": hl.or_missing(
+                bin_ht[f"_filter_{bin_id}"],
+                hl.scan.count_where(bin_ht[f"_filter_{bin_id}"]),
             )
-            for rname in bin_stats
+            for bin_id in bin_expr
         }
     )
+    bin_ht = bin_ht.key_by("locus", "alleles")
 
+    # Annotate globals with variant counts per group defined by bin_expr. This is used to determine bin assignment
     bin_ht = bin_ht.annotate_globals(
-        bin_stats=hl.literal(
-            bin_stats,
-            dtype=hl.tstruct(
+        bin_group_variant_counts=bin_ht.aggregate(
+            hl.Struct(
                 **{
-                    bin_id: hl.tstruct(
-                        n=hl.tint64,
-                        quantiles=hl.tarray(hl.tfloat64),
-                        bin_boundaries=hl.tarray(hl.tfloat64),
-                        global_bin_indices=hl.tarray(hl.tint32),
-                        merged_bins=hl.tdict(hl.tint32, hl.tint32),
-                    )
+                    bin_id: hl.agg.filter(bin_ht[f"_filter_{bin_id}"], hl.agg.count(),)
                     for bin_id in bin_expr
                 }
-            ),
-        )
-    )
-
-    # Annotate the bin as the index in the unique boundaries array
-    bin_ht = bin_ht.annotate(
-        **{
-            bin_id: hl.or_missing(
-                bin_ht[f"_filter_{bin_id}"],
-                hl.binary_search(
-                    bin_ht.bin_stats[bin_id].bin_boundaries, bin_ht._score
-                ),
             )
-            for bin_id in bin_expr
-        }
+        )
     )
 
-    # Convert the bin to global bin by expanding merged bins, that is:
-    # If a value falls in a bin that needs expansion, assign it randomly to one of the expanded bins
-    # Otherwise, simply modify the bin to its global index (with expanded bins that is)
+    logger.info(f"Binning ranked rows into {n_bins}...")
     bin_ht = bin_ht.select(
         "snv",
         **{
-            bin_id: hl.if_else(
-                bin_ht.bin_stats[bin_id].merged_bins.contains(bin_ht[bin_id]),
-                bin_ht.bin_stats[bin_id].global_bin_indices[bin_ht[bin_id]]
-                + hl.int(
-                    hl.rand_unif(
-                        0, bin_ht.bin_stats[bin_id].merged_bins[bin_ht[bin_id]] + 1
+            bin_id: hl.int(
+                hl.floor(
+                    (
+                        n_bins
+                        * (
+                            bin_ht[f"{bin_id}_rank"]
```

Comment:
There are actually 101 bins the way the code is written, because the largest rank (say it's 1002 in a given grouping) will give a value of 1 * n_bins = 100 (because the rank 1002 will be divided by the total number of variants in the grouping, 1002). Basically, you're starting a new bin for the single final variant in each grouping.

I think if you want to only have 100 bins, you'll need to subtract 1 off the `{bin_id}_rank` in line 121 before dividing by `hl.float64(bin_ht.bin_group_variant_counts[bin_id]` on line 122. This will handle the other end (the first variant in the grouping) fine, because `hl.floor()` will still return zero for the first variant.

Reply:
I'm not sure I agree, the rank is actually 0 to (n_variants)

Notebook test:
[cleanup_new_binning_function.html.zip](https://github.com/broadinstitute/gnomad_methods/files/5895478/cleanup_new_binning_function.html.zip)

Reply:
Can you show what the `{bin_id}_rank` is in the notebook, and not just the final bin assignment? I agree if the rank is 0 to (n_variants - 1) then the code as written is fine (basically the ranking already has 1 subtracted off it). It's just that I assumed `hl.scan.count_where(bin_ht[f"_filter_{bin_id}"])` would be an actual count, which starts a 1 and not 0.

Reply:
<img width="278" alt="Screen Shot 2021-01-29 at 2 09 05 PM" src="https://user-images.githubusercontent.com/33063077/106327597-a70a6180-623b-11eb-95aa-6db8747a4f0f.png">
<img width="268" alt="Screen Shot 2021-01-29 at 2 09 12 PM" src="https://user-images.githubusercontent.com/33063077/106327603-a8d42500-623b-11eb-8a3b-99705f9768d5.png">

Reply:
I thought it was too at first, but got the wrong binning when I tested so changed it

Reply:
Thanks! I wonder if the behavior changes when it's a `scan` -- I know the sum mechanism for `hl.scan()` doesn't count the current row, so maybe that's what's happening here.

In any case, the code is working as intended, so no changes required.